### PR TITLE
updated Kubeflow roadmap 

### PR DESCRIPTION
**Pull request by jbottum**
updating the Kubeflow roadmap, per issue  #4903

### CODE REVIEW

File: ROADMAP.md [hunk 1]
```diff
@@ -114,37 +114,39 @@ Support for hardware accelerated training and inference
...
-
 * ISTIO will be required as a service mesh and for AuthN and AuthZ support
 * We will only support a single shared Kubeflow deployment per Kubernetes cluster
-  * Users can consume Kubeflow in their own, isolated namespace
-* Upgrades will require downtime
-* Upgradability will not support advanced customization (e.g. custom overlays) for Kustomize packages
+Users can consume Kubeflow in their own, isolated namespace
+Upgrades will require downtime
+
+## Kubeflow 1.1 Features, Target release: June 2020
```
Comment (Jeffwan) [inline ROADMAP.md:144]: I miss this week's community meeting. Did community have a discussion on 1.1 features?

Comment (jbottum) [reply ROADMAP.md:144]: @Jeffwan Hey Jeff, I am not sure if we discussed the features in the PM call or the Community Meeting.

Comment (Jeffwan) [reply ROADMAP.md:144]: No, Just think this is highlights and doesn't come with details. Wondering if we have more details somewhere

**Comment by jlewi** [comment]
@animeshsingh @jinchihe Anything for fairing we should include here?

@gaocegege @johnugeorge @hougangliu  Anything for Katib? Does Katib have an updated ROADMAP?

@kimwnasptd Anything for notebooks?

@jessiezcc Anything for pipelines?

@johnugeorge @yanniszark Anything for onprem?

@animeshsingh @cliveseldon Anything for KFServing? Is the KFServing ROADMAP up to date?

**Comment by jbottum** [comment]
@jlewi Hey Jeremy, In the Kubeflow 1.0 webinar, we will promote that publishing a roadmap is part of our stable processes.   Could we please push this through before Tuesday

**Comment by jlewi** [comment]
Are we doing the ROADMAP this week or next week? 

I would really like to get input from the various area/owners. The area/owners are the ones responsible for committing and driving the work. So I think its important we have buy in from the area/owners.

**Comment by jbottum** [comment]
@jlewi hey Jeremy - I would like to get this updated before the webinar.    It says that we are missing a CLA, but I checked mine and it says that I have already signed.   Any ideas ?

**Comment by Jeffwan** [comment]
@jbottum It seems your commit author doesn't match with your github user. Try to edit the author like this

```
git commit --amend --author="Josh Bottum <joshbottum@gmail.com>" --no-edit
```

**Comment by jbottum** [comment]
@abhi-g @richardsliu Hey Abhishek, Hey Richard - Would you guys please approve this update to our roadmap.   I am sure we can add more improvements later but it pretty far out of date.   Thanks.

**Comment by abhi-g** [comment]
I think the changes in Kubeflow repo are minimal and roadmaps for features
in individual components should be in the corresponding repositories. You
might want to just add pointers to those. For example, fairing, xgboost
operator, kfserving all have updated roadmaps.

For Kubeflow/Kubeflow Jeremy has plans on Kustomize changes to make easier
upgrades preserving custom environment setup of users. Again that may
belong in manifests repo, but worth a mention here.

To @jlewi's point it's not a good idea to make commitments in roadmap that
those individual components' owners may not be onboard with or haven't
committed yet.

File: ROADMAP.md [hunk 2]
```diff
@@ -114,37 +114,39 @@ Support for hardware accelerated training and inference
...
 * Profile controller and UI for multiuser management
 
-The following applications will likely be of beta quality for the 1.0 release with a goal of graduating to 1.0 in Q1 2020
-
+The following applications are considered in a beta version in Kubeflow 1.0.
 * Katib for hyper-parameter tuning
-* fairing SDK to facilite use of notebooks for build-train-deploy
-* Metadata SDK, UI, and backend
+* fairing SDK to facilitate use of notebooks for build-train-deploy
+* Kale which extends jupyter notebooks to create, create, run, explore KF pipelines
```
Comment (rmbrad) [inline ROADMAP.md:133]: Create repeated

```suggestion
* Kale which extends jupyter notebooks to create, run, and explore KF pipelines
```

**Comment by jbottum** [comment]
@googlebot I signed it!

**Comment by yanniszark** [comment]
@jlewi I think the ROADMAP should involve multitenancy and a clear plan on how Kubeflow would fit into existing mechanisms.
This is related to https://groups.google.com/forum/#!topic/kubeflow-discuss/whS_j4RlEJo, https://github.com/kubeflow/kubeflow/issues/4899 and discussions about SubjectAccessReview, application initialization, etc.

This will:
1. Help Kubeflow admins easily reason about how to give permissions to users.
2. Enable integrations of existing tools for k8s administration.
3. Pave a clear path and best-practice for applications to integrate with Kubeflow.

**Comment by kimwnasptd** [comment]
@jlewi Also regarding the notebooks I would like to add support for start/stopping Notebook Servers for 1.1.

This way users won't have to delete their Notebooks to prevent them from reserving cloud resources even when they are idle.

**Comment by jbottum** [comment]
@googlebot I signed it!

**Comment by jbottum** [comment]
@googlebot I fixed it.

File: ROADMAP.md [hunk 3]
```diff
@@ -114,37 +114,38 @@ Support for hardware accelerated training and inference
...
+Upgrades will require downtime
+
+## Kubeflow 1.1 Features, Target release: June 2020
+
+Initial features under review include:
 
-  * We expect application to expose a list of parameters that can be customized without breaking upgradability
-  * Customization beyond these parameters will require manual configuration on upgrade
+* Improved upgradability functionality
+* Enhanced security use cases for GCP users
```
Comment (jlewi) [inline ROADMAP.md:149]: I think we can be more specific. We are planning on adding support for private GKE and Anthos Service Mesh link (https://cloud.google.com/service-mesh/docs); The tracking issue is kubeflow/website/issues/1705

File: ROADMAP.md [hunk 4]
```diff
@@ -114,37 +114,38 @@ Support for hardware accelerated training and inference
...
-
 * ISTIO will be required as a service mesh and for AuthN and AuthZ support
 * We will only support a single shared Kubeflow deployment per Kubernetes cluster
-  * Users can consume Kubeflow in their own, isolated namespace
-* Upgrades will require downtime
-* Upgradability will not support advanced customization (e.g. custom overlays) for Kustomize packages
+Users can consume Kubeflow in their own, isolated namespace
+Upgrades will require downtime
+
+## Kubeflow 1.1 Features, Target release: June 2020
```
Comment (jlewi) [inline ROADMAP.md:144]: @jbottum Could you provide a link to the kanban board for 1.1?
https://github.com/orgs/kubeflow/projects/36

File: ROADMAP.md [hunk 5]
```diff
@@ -114,37 +114,38 @@ Support for hardware accelerated training and inference
...
+
+Initial features under review include:
 
-  * We expect application to expose a list of parameters that can be customized without breaking upgradability
-  * Customization beyond these parameters will require manual configuration on upgrade
+* Improved upgradability functionality
```
Comment (jlewi) [inline ROADMAP.md:148]: For improved upgrade functionality I think you can link to kubeflow/kfctl#304

File: ROADMAP.md [hunk 6]
```diff
@@ -114,37 +114,38 @@ Support for hardware accelerated training and inference
...
+* Enhanced security use cases for GCP users
+* CVE scanning and mitigation
```
Comment (jlewi) [inline ROADMAP.md:150]: Do we have an issue tracking this? @jtfogarty can we nail down what we plan on delivering in the 1.1. release?

File: ROADMAP.md [hunk 7]
```diff
@@ -114,37 +114,38 @@ Support for hardware accelerated training and inference
...
+* Improved workflow automation tools (fairing and kale) to simplify and mature the Core and EcoSystem supported CUJs
```
Comment (jlewi) [inline ROADMAP.md:151]: Can we be more specific about what we plan on delivering? Can we link to issues or Kanban boards.

**Comment by jlewi** [comment]
> @jlewi I think the ROADMAP should involve multitenancy and a clear plan on how Kubeflow would fit into existing mechanisms.

   * @yanniszark Can you suggest some deliverables for Kubeflow 1.1? 

@jbottum Can we add more detail to the roadmap? e.g. @kimwnasptd 's item about starting stopping notebook?

@kimwnasptd what about a Tensorboard manager? I'm guessing that won't land in 1.1; especially if we are doing it as a gsoc project.

Can we come up with 2-3 themes for the 1.1 release?

I think enterprise readiness could be one. 

For 2020 I think we need some high-level way to communicate and orient the various work streams. In the [past](https://docs.google.com/presentation/d/1gdGjRd2u2tIL691VI3VaYKOjYIy0plDpGZeEM-GK3CY/edit#slide=id.g7c38e4c3ec_0_261) we'd describe the work as individual apps, then connecting apps, and then productionisation and hardening.

I think another way to think about it would be to borrow the concepts of [Day 0, Day 1, Day 2](https://codilime.com/day-0-day-1-day-2-the-software-lifecycle-in-the-cloud-age/) from DevOps.

When it comes to Kubeflow, I think we are still on Day 0. Right now customers can

* Deploy Kubeflow
* Run the examples
* Start to build and deploy their own models

I think we are still on Day 0 though because I think there is still a huge gap to actually deploying an application and demonstrating that it actually improves KPIs. 

We probably need to come up with some suitable definitions for Day 1 and Day 2 in the context of ML. Maybe someone already has?

My initial thought would be something like the following

* Day 0 - Identify the problem, data, and do some exploration to see whether the problem is amenable to ML
* Day 1 - Get an MVP into production and demonstrate positive impact on KPIs
* Day 2 - 
  * Iterate on the model (e.g. replace linear model with deep model), 
  * continuously retrain and deploy
  * Deal with ML outages (e.g non-stationary data, skew?)

Here are some references that might be useful

* [MLOps: Continuous delivery and automation pipelines in machine learning](https://cloud.google.com/solutions/machine-learning/mlops-continuous-delivery-and-automation-pipelines-in-machine-learning) - This doc defines 3 levels so maybe we borrow those?
* [MLOps Sig Roadmap](https://github.com/cdfoundation/sig-mlops/blob/master/roadmap/2020/MLOpsRoadmap2020.md)

As an example, if I look at our [mnist example](https://github.com/kubeflow/examples/tree/master/mnist) we don't have a story for how people would measure the performance of the model in the sample application.







.

**Comment by yanniszark** [comment]
> @yanniszark Can you suggest some deliverables for Kubeflow 1.1?

Sure, some things I can think of:
* Establishment of Kubeflow Policy / Guidelines on how to implement authorization in web applications. Propose SubjectAccessReview in order to use K8s RBAC as the source of truth for Authz.
* Guidelines on how cluster admins can interact with Kubeflow's authorization. There are already some difficulties with the self-serve model, process of adding contributors to a namespace and the way KFAM is using magic annotations on RoleBindings (https://github.com/kubeflow/kubeflow/issues/4574 https://github.com/kubeflow/kubeflow/issues/4889 https://github.com/kubeflow/kubeflow/issues/4936 https://github.com/kubeflow/kubeflow/issues/4924 https://github.com/kubeflow/kubeflow/issues/4938). Document current workarounds.
* Decide when the CentralDashboard should show a namespace. Right now, this is done with KFAM in an error-prone way (magic annotations on RoleBindings). Design doc exploring different options (change KFAM to use SubjectAccessReview, use a model like GCP Console checking read permission on namespace, etc.). This is also related to the 2nd bullet.
* Ability to turn off the self-serve mode, as in many environments there are mechanisms other than the Kubeflow Dashboard that provision/share an environment for/with the user. This has popped up a couple of times in Slack and we have needed it as well.

@jlewi what do you think? I can open issues and start working on design if there's agreement.

**Comment by thesuperzapper** [comment]
> @jlewi Also regarding the notebooks I would like to add support for start/stopping Notebook Servers for 1.1.
> 
> This way users won't have to delete their Notebooks to prevent them from reserving cloud resources even when they are idle.

I made an issue to track this #4951

**Comment by jlewi** [comment]
@yanniszark  LGTM

**Comment by jlewi** [comment]
@jbottum do you want to incorporate this feedback into the PR?

**Comment by yanniszark** [comment]
@jlewi thanks, I edited my original comment with the newly opened issues.

**Comment by thesuperzapper** [comment]
I have added some feature issues I think will make Kubeflow more 'enterprise-ready', I would love to see them make it into a coming release:
* #4950 - PVC discovery
* #4951 - Stop/Pause Notebook Servers
* #4952 - Allow creation of profiles within dashboard
* #4953 - Enforce Profile/Namespace naming in registration flow

**Comment by jlewi** [comment]
@jbottum do you want to update your PR to include the feedback on the thread? Also if you rebase and squash commits that will fix the CLA check.

**Pull request merged**